{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "831bc6be-9360-4164-8750-6833d506098c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/derbhladuffy/miniconda3/envs/py310_env/bin/python'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.executable\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8087f35f-82d9-4b8a-bb13-137e1f4e35d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: requests in ./lib/python3.10/site-packages (2.32.5)\n",
      "Requirement already satisfied: beautifulsoup4 in ./lib/python3.10/site-packages (4.13.5)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in ./lib/python3.10/site-packages (from requests) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in ./lib/python3.10/site-packages (from requests) (3.11)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in ./lib/python3.10/site-packages (from requests) (2.5.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in ./lib/python3.10/site-packages (from requests) (2025.10.5)\n",
      "Requirement already satisfied: soupsieve>1.2 in ./lib/python3.10/site-packages (from beautifulsoup4) (2.5)\n",
      "Requirement already satisfied: typing-extensions>=4.0.0 in ./lib/python3.10/site-packages (from beautifulsoup4) (4.15.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install requests beautifulsoup4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62c8b02d-1fe8-49b3-aa36-08ff28a9182c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /Users/derbhladuffy/Desktop/diss_project/annaconda_env_colabfold\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Your chosen working directory\n",
    "output_dir = \"/Users/derbhladuffy/Desktop/diss_project/annaconda_env_colabfold\"\n",
    "\n",
    "# Create it if it doesn't exist\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Switch notebook working directory to it\n",
    "os.chdir(output_dir)\n",
    "\n",
    "print(\"Current working directory:\", os.getcwd())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd2d0251-83cc-486c-8c66-f52026a5348f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## try running a script to one 50aa sequence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cc7229fe-aead-4684-81a5-8b046a451336",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html>\n",
      "<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en-US\" prefix=\"bioschema: https://bioschemas.org/\">\n",
      "<head>\n",
      "    <meta http-equiv=\"Content-Type\" content=\"text/html;charset=iso-8859-1\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"application/ld+json\">\n",
      "        {\n",
      "            \"@context\": [ \"https://schema.org/\" ],\n",
      "            \"@type\": \"Organization\",\n",
      "            \"url\": \"https://web.expasy.org/randseq\",\n",
      "            \"name\": \"RandSeq\",\n",
      "            \"logo\": \"https://web.expasy.org/randseq/images/expasy_logos/Expasy_logo_base_addon.svg\",\n",
      "            \"memberof\": {\n",
      "                \"@type\": \"NGO\",\n",
      "                \"name\": \"SIB Swiss Institute of Bioinformatics\",\n",
      "                \"url\": \"https://sib.swiss/\"\n",
      "            },\n",
      "            \"parentOrganization\": {\n",
      "                \"@type\": \"Organization\",\n",
      "                \"name\": \"Expasy\",\n",
      "                \"url\": \"https://www.expasy.org/\"\n",
      "            }\n",
      "        }\n",
      "    </script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <link rel=\"stylesheet\" href=\"/css/normalize.css?v=1.1\" type=\"text/css\" media=\"screen\" charset=\"utf-8\">\n",
      "    <link rel=\"stylesheet\" href=\"/css/expasy_geneva.css?v=1.1\" type=\"text/css\" charset=\"utf-8\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/jquery.min.js\"></script>\n",
      "    <script type=\"text/javascript\" src=\"/js/main.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/sp_tools/matomo-randseq-tracking.js\"></script>\n",
      "    <script type=\"text/javascript\" defer=\"defer\" src=\"https://matomo.sib.swiss/matomo.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "\n",
      "\t<title>Random entry: RND237411</title>\n",
      "</head>\n",
      "<body  typeof=\"schema:WebPage\"> \n",
      "<!-- start of header -->\n",
      "<header class=\"header-grid\">\n",
      "    <div class=\"header-elmt\">\n",
      "        <img alt=\"Expasy logo\" src=\"/images/expasy_logos/Expasy_logo_base_addon.svg\">\n",
      "    </div>\n",
      "    <h4 class=\"header-elmt\"><a href=\"https://web.expasy.org/randseq\">RandSeq</a></h4>\n",
      "    <nav class=\"header-elmt\" itemscope itemtype=\"http://www.schema.org/SiteNavigationElement\">\n",
      "        <ul class=\"menu-2-items\">\n",
      "            <li>\n",
      "                <img src=\"/images/icons/home.svg\" alt=\"home\">\n",
      "                <a itemprop=\"url\" href=\"/randseq/\">\n",
      "                    <span itemprop=\"name\">Home</span>\n",
      "                </a>\n",
      "            </li>\n",
      "            <li>\n",
      "                <img src=\"/images/icons/contact.svg\" alt=\"contact\">\n",
      "                <a itemprop=\"url\" href=\"/contact\">\n",
      "                    <span itemprop=\"name\">Contact</span>\n",
      "                </a>\n",
      "            </li>\n",
      "        </ul>\n",
      "    </nav>\n",
      "</header>\n",
      "<!-- end of header, start of main -->\n",
      "<main class=\"main-grid\">    <!-- automatically added by Apache -->\n",
      "<h1 class=\"main-elmt\">Random entry: RND237411</h1>\n",
      "<h3 class=\"main-elmt\">Entry in UniProtKB Swiss-Prot format</h3>\n",
      "<pre class=\"main-elmt\">\n",
      "\n",
      "ID   RND_237411              Unreviewed;        50 AA.\n",
      "AC   RND237411;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; 01B42E20EBCE56C3 CRC64;\n",
      "     ENIETSLKLD GDSAALGLVG SRKVAYWEFG IPHTLVNQKI TMDQRRFACP \n",
      "//\n",
      "</pre>\n",
      "<div class=\"main-elmt\">\n",
      "<hr><br><h3>Submit Entry to Expasy tools</h3>\n",
      "<br>\n",
      "<table class=\"type-1\" align=\"center\" width=\"100%\">\n",
      "<tr>\n",
      "<td><img width=\"100 rem\" src=\"/images/prosite/prosite_trans.png\"></td>\n",
      "<td><img width=\"60 rem\" src=\"/images/others/tools50x50.gif\" style=\"display: inline;\" align=\"center\">&nbsp;&nbsp;<strong>Sequence analysis tools</strong></td>\n",
      "</tr>\n",
      "<tr>\n",
      "<td><a href=\"https://prosite.expasy.org/scanprosite/?RND237411\">ScanProsite</a></td>\n",
      "<td>\n",
      "<a href=\"/cgi-bin/compute_pi/pi_tool.cgi?RND237411\">Compute pI/mW</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_cutter/peptidecutter.pl?RND237411\">PeptideCutter</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_mass/peptide-mass.pl?RND237411\">PeptideMass</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protparam/protparam.cgi?RND237411\">ProtParam</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protscale/protscale.pl?RND237411\">ProtScale</a>\n",
      "</td>\n",
      "</tr>\n",
      "</table>\n",
      "</div>\n",
      "    </main>\n",
      "    <!-- end of main, start of footer -->\n",
      "    <footer class=\"footer-grid\">\n",
      "\t    <div class=\"footer-elmt\">\n",
      "            <img alt=\"SIB logo\" src=\"/images/sib_images/sib-mark.svg\">\n",
      "\t\t</div>\n",
      "\t\t<div class=\"footer-elmt\" property=\"schema:creator\" typeof=\"schema:NGO\">\n",
      "           <a href=\"https://www.expasy.org\" target=\"_blank\">Expasy</a> \n",
      "\t\t   is operated by the <a property=\"schema:url\" href=\"https://sib.swiss\" target=\"_blank\">SIB Swiss Institute of Bioinformatics</a>\n",
      "\t\t   <br>\n",
      "\t\t   <a href=\"https://www.expasy.org/terms-of-use\" target=\"_blank\">Terms of Use</a>\n",
      "           |\n",
      "           <a href=\"https://www.sib.swiss/privacy-policy\" target=\"_blank\">Privacy policy</a>\n",
      "\t    </div>\n",
      "\t    <div class=\"footer-elmt\">\n",
      "\t        <a href=\"#top\">Back to the top</a>\n",
      "\t    </div>\n",
      "    </footer>\n",
      "    <!-- end of footer -->\n",
      "</body>\n",
      "</html>\n",
      "ID   RND_237411              Unreviewed;        50 AA.\n",
      "AC   RND237411;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; 01B42E20EBCE56C3 CRC64;\n",
      "     ENIETSLKLD GDSAALGLVG SRKVAYWEFG IPHTLVNQKI TMDQRRFACP \n",
      "//\n"
     ]
    }
   ],
   "source": [
    "##Contacts ExPASy, requests a random protein sequence, and returns the Swiss-Prot text entry:\n",
    "\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "def generate_randseq(length=50):                            # ðŸ”¶ CHANGE THIS (sequence length)\n",
    "    url = \"https://web.expasy.org/cgi-bin/randseq/randseq.pl\"\n",
    "    data = {\n",
    "        \"size\": str(length),                                # ðŸ”¶ Set by the length argument above\n",
    "        \"composition\": \"average\"                            # ðŸ”¶ CHANGE THIS if you want equal/user-defined/specific sequence\n",
    "    }\n",
    "\n",
    "    r = requests.post(url, data=data)\n",
    "\n",
    "    # TEMPORARY: print response\n",
    "    print(r.text)\n",
    "\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "    seq_block = soup.find(\"pre\")\n",
    "\n",
    "    if not seq_block:\n",
    "        print(\"Could not find sequence in response.\")\n",
    "        return None\n",
    "\n",
    "    return seq_block.text.strip()\n",
    "\n",
    "test_seq = generate_randseq(50)\n",
    "print(test_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "680a478d-31c3-49a7-9e5f-4d02a178fff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Takes the Swiss-Prot text, extracts the amino-acid sequence + accession, and saves a clean FASTA file.\n",
    "\n",
    "def swissprot_to_fasta(text, outname=\"test_randseq_1.fasta\"):            # ðŸ”¶ CHANGE THIS (output FASTA filename)\n",
    "    # Split into lines\n",
    "    lines = text.splitlines()\n",
    "\n",
    "    # Extract accession number (AC   RNDxxxxx;)\n",
    "    ac_line = [l for l in lines if l.startswith(\"AC\")][0]\n",
    "    accession = ac_line.split()[1].replace(\";\", \"\")\n",
    "\n",
    "    # Extract sequence lines between SQ and //\n",
    "    seq_lines = []\n",
    "    capture = False\n",
    "    for line in lines:\n",
    "        if line.startswith(\"SQ\"):\n",
    "            capture = True\n",
    "            continue\n",
    "        if capture:\n",
    "            if line.startswith(\"//\"):\n",
    "                break\n",
    "            seq_lines.append(line.strip())\n",
    "\n",
    "    # Join and remove spaces\n",
    "    sequence = \"\".join(seq_lines).replace(\" \", \"\")\n",
    "\n",
    "    # Build FASTA string\n",
    "    fasta = f\">{accession}\\n{sequence}\"\n",
    "\n",
    "    # Save it\n",
    "    with open(outname, \"w\") as f:\n",
    "        f.write(fasta + \"\\n\")\n",
    "\n",
    "    print(f\"Saved FASTA as {outname}\")\n",
    "    return fasta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2fa8008f-e2de-4c93-9ffb-93333ab68126",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html>\n",
      "<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en-US\" prefix=\"bioschema: https://bioschemas.org/\">\n",
      "<head>\n",
      "    <meta http-equiv=\"Content-Type\" content=\"text/html;charset=iso-8859-1\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"application/ld+json\">\n",
      "        {\n",
      "            \"@context\": [ \"https://schema.org/\" ],\n",
      "            \"@type\": \"Organization\",\n",
      "            \"url\": \"https://web.expasy.org/randseq\",\n",
      "            \"name\": \"RandSeq\",\n",
      "            \"logo\": \"https://web.expasy.org/randseq/images/expasy_logos/Expasy_logo_base_addon.svg\",\n",
      "            \"memberof\": {\n",
      "                \"@type\": \"NGO\",\n",
      "                \"name\": \"SIB Swiss Institute of Bioinformatics\",\n",
      "                \"url\": \"https://sib.swiss/\"\n",
      "            },\n",
      "            \"parentOrganization\": {\n",
      "                \"@type\": \"Organization\",\n",
      "                \"name\": \"Expasy\",\n",
      "                \"url\": \"https://www.expasy.org/\"\n",
      "            }\n",
      "        }\n",
      "    </script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <link rel=\"stylesheet\" href=\"/css/normalize.css?v=1.1\" type=\"text/css\" media=\"screen\" charset=\"utf-8\">\n",
      "    <link rel=\"stylesheet\" href=\"/css/expasy_geneva.css?v=1.1\" type=\"text/css\" charset=\"utf-8\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/jquery.min.js\"></script>\n",
      "    <script type=\"text/javascript\" src=\"/js/main.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/sp_tools/matomo-randseq-tracking.js\"></script>\n",
      "    <script type=\"text/javascript\" defer=\"defer\" src=\"https://matomo.sib.swiss/matomo.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "\n",
      "\t<title>Random entry: RND255565</title>\n",
      "</head>\n",
      "<body  typeof=\"schema:WebPage\"> \n",
      "<!-- start of header -->\n",
      "<header class=\"header-grid\">\n",
      "    <div class=\"header-elmt\">\n",
      "        <img alt=\"Expasy logo\" src=\"/images/expasy_logos/Expasy_logo_base_addon.svg\">\n",
      "    </div>\n",
      "    <h4 class=\"header-elmt\"><a href=\"https://web.expasy.org/randseq\">RandSeq</a></h4>\n",
      "    <nav class=\"header-elmt\" itemscope itemtype=\"http://www.schema.org/SiteNavigationElement\">\n",
      "        <ul class=\"menu-2-items\">\n",
      "            <li>\n",
      "                <img src=\"/images/icons/home.svg\" alt=\"home\">\n",
      "                <a itemprop=\"url\" href=\"/randseq/\">\n",
      "                    <span itemprop=\"name\">Home</span>\n",
      "                </a>\n",
      "            </li>\n",
      "            <li>\n",
      "                <img src=\"/images/icons/contact.svg\" alt=\"contact\">\n",
      "                <a itemprop=\"url\" href=\"/contact\">\n",
      "                    <span itemprop=\"name\">Contact</span>\n",
      "                </a>\n",
      "            </li>\n",
      "        </ul>\n",
      "    </nav>\n",
      "</header>\n",
      "<!-- end of header, start of main -->\n",
      "<main class=\"main-grid\">    <!-- automatically added by Apache -->\n",
      "<h1 class=\"main-elmt\">Random entry: RND255565</h1>\n",
      "<h3 class=\"main-elmt\">Entry in UniProtKB Swiss-Prot format</h3>\n",
      "<pre class=\"main-elmt\">\n",
      "\n",
      "ID   RND_255565              Unreviewed;        50 AA.\n",
      "AC   RND255565;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; 643533CDFED9590C CRC64;\n",
      "     IGLEGWHDKI ILTGMRLRPA ECVESAKYVQ TVNSTNDDFP GLARSKQLAF \n",
      "//\n",
      "</pre>\n",
      "<div class=\"main-elmt\">\n",
      "<hr><br><h3>Submit Entry to Expasy tools</h3>\n",
      "<br>\n",
      "<table class=\"type-1\" align=\"center\" width=\"100%\">\n",
      "<tr>\n",
      "<td><img width=\"100 rem\" src=\"/images/prosite/prosite_trans.png\"></td>\n",
      "<td><img width=\"60 rem\" src=\"/images/others/tools50x50.gif\" style=\"display: inline;\" align=\"center\">&nbsp;&nbsp;<strong>Sequence analysis tools</strong></td>\n",
      "</tr>\n",
      "<tr>\n",
      "<td><a href=\"https://prosite.expasy.org/scanprosite/?RND255565\">ScanProsite</a></td>\n",
      "<td>\n",
      "<a href=\"/cgi-bin/compute_pi/pi_tool.cgi?RND255565\">Compute pI/mW</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_cutter/peptidecutter.pl?RND255565\">PeptideCutter</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_mass/peptide-mass.pl?RND255565\">PeptideMass</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protparam/protparam.cgi?RND255565\">ProtParam</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protscale/protscale.pl?RND255565\">ProtScale</a>\n",
      "</td>\n",
      "</tr>\n",
      "</table>\n",
      "</div>\n",
      "    </main>\n",
      "    <!-- end of main, start of footer -->\n",
      "    <footer class=\"footer-grid\">\n",
      "\t    <div class=\"footer-elmt\">\n",
      "            <img alt=\"SIB logo\" src=\"/images/sib_images/sib-mark.svg\">\n",
      "\t\t</div>\n",
      "\t\t<div class=\"footer-elmt\" property=\"schema:creator\" typeof=\"schema:NGO\">\n",
      "           <a href=\"https://www.expasy.org\" target=\"_blank\">Expasy</a> \n",
      "\t\t   is operated by the <a property=\"schema:url\" href=\"https://sib.swiss\" target=\"_blank\">SIB Swiss Institute of Bioinformatics</a>\n",
      "\t\t   <br>\n",
      "\t\t   <a href=\"https://www.expasy.org/terms-of-use\" target=\"_blank\">Terms of Use</a>\n",
      "           |\n",
      "           <a href=\"https://www.sib.swiss/privacy-policy\" target=\"_blank\">Privacy policy</a>\n",
      "\t    </div>\n",
      "\t    <div class=\"footer-elmt\">\n",
      "\t        <a href=\"#top\">Back to the top</a>\n",
      "\t    </div>\n",
      "    </footer>\n",
      "    <!-- end of footer -->\n",
      "</body>\n",
      "</html>\n",
      "ID   RND_255565              Unreviewed;        50 AA.\n",
      "AC   RND255565;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; 643533CDFED9590C CRC64;\n",
      "     IGLEGWHDKI ILTGMRLRPA ECVESAKYVQ TVNSTNDDFP GLARSKQLAF \n",
      "//\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5f33eae0-b774-4093-a8c2-54848f269ebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html>\n",
      "<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en-US\" prefix=\"bioschema: https://bioschemas.org/\">\n",
      "<head>\n",
      "    <meta http-equiv=\"Content-Type\" content=\"text/html;charset=iso-8859-1\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"application/ld+json\">\n",
      "        {\n",
      "            \"@context\": [ \"https://schema.org/\" ],\n",
      "            \"@type\": \"Organization\",\n",
      "            \"url\": \"https://web.expasy.org/randseq\",\n",
      "            \"name\": \"RandSeq\",\n",
      "            \"logo\": \"https://web.expasy.org/randseq/images/expasy_logos/Expasy_logo_base_addon.svg\",\n",
      "            \"memberof\": {\n",
      "                \"@type\": \"NGO\",\n",
      "                \"name\": \"SIB Swiss Institute of Bioinformatics\",\n",
      "                \"url\": \"https://sib.swiss/\"\n",
      "            },\n",
      "            \"parentOrganization\": {\n",
      "                \"@type\": \"Organization\",\n",
      "                \"name\": \"Expasy\",\n",
      "                \"url\": \"https://www.expasy.org/\"\n",
      "            }\n",
      "        }\n",
      "    </script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <link rel=\"stylesheet\" href=\"/css/normalize.css?v=1.1\" type=\"text/css\" media=\"screen\" charset=\"utf-8\">\n",
      "    <link rel=\"stylesheet\" href=\"/css/expasy_geneva.css?v=1.1\" type=\"text/css\" charset=\"utf-8\">\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/jquery.min.js\"></script>\n",
      "    <script type=\"text/javascript\" src=\"/js/main.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "    <script type=\"text/javascript\" src=\"/js/sp_tools/matomo-randseq-tracking.js\"></script>\n",
      "    <script type=\"text/javascript\" defer=\"defer\" src=\"https://matomo.sib.swiss/matomo.js\"></script>\n",
      "    <!-- automatically added by Apache -->\n",
      "\n",
      "\t<title>Random entry: RND454178</title>\n",
      "</head>\n",
      "<body  typeof=\"schema:WebPage\"> \n",
      "<!-- start of header -->\n",
      "<header class=\"header-grid\">\n",
      "    <div class=\"header-elmt\">\n",
      "        <img alt=\"Expasy logo\" src=\"/images/expasy_logos/Expasy_logo_base_addon.svg\">\n",
      "    </div>\n",
      "    <h4 class=\"header-elmt\"><a href=\"https://web.expasy.org/randseq\">RandSeq</a></h4>\n",
      "    <nav class=\"header-elmt\" itemscope itemtype=\"http://www.schema.org/SiteNavigationElement\">\n",
      "        <ul class=\"menu-2-items\">\n",
      "            <li>\n",
      "                <img src=\"/images/icons/home.svg\" alt=\"home\">\n",
      "                <a itemprop=\"url\" href=\"/randseq/\">\n",
      "                    <span itemprop=\"name\">Home</span>\n",
      "                </a>\n",
      "            </li>\n",
      "            <li>\n",
      "                <img src=\"/images/icons/contact.svg\" alt=\"contact\">\n",
      "                <a itemprop=\"url\" href=\"/contact\">\n",
      "                    <span itemprop=\"name\">Contact</span>\n",
      "                </a>\n",
      "            </li>\n",
      "        </ul>\n",
      "    </nav>\n",
      "</header>\n",
      "<!-- end of header, start of main -->\n",
      "<main class=\"main-grid\">    <!-- automatically added by Apache -->\n",
      "<h1 class=\"main-elmt\">Random entry: RND454178</h1>\n",
      "<h3 class=\"main-elmt\">Entry in UniProtKB Swiss-Prot format</h3>\n",
      "<pre class=\"main-elmt\">\n",
      "\n",
      "ID   RND_454178              Unreviewed;        50 AA.\n",
      "AC   RND454178;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; F6D65D5CD1CDBA1B CRC64;\n",
      "     LAFSVYKTIG RLDPNELVWD ISGGHICMAQ PNEKTQAERD SGTLLKRAFV \n",
      "//\n",
      "</pre>\n",
      "<div class=\"main-elmt\">\n",
      "<hr><br><h3>Submit Entry to Expasy tools</h3>\n",
      "<br>\n",
      "<table class=\"type-1\" align=\"center\" width=\"100%\">\n",
      "<tr>\n",
      "<td><img width=\"100 rem\" src=\"/images/prosite/prosite_trans.png\"></td>\n",
      "<td><img width=\"60 rem\" src=\"/images/others/tools50x50.gif\" style=\"display: inline;\" align=\"center\">&nbsp;&nbsp;<strong>Sequence analysis tools</strong></td>\n",
      "</tr>\n",
      "<tr>\n",
      "<td><a href=\"https://prosite.expasy.org/scanprosite/?RND454178\">ScanProsite</a></td>\n",
      "<td>\n",
      "<a href=\"/cgi-bin/compute_pi/pi_tool.cgi?RND454178\">Compute pI/mW</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_cutter/peptidecutter.pl?RND454178\">PeptideCutter</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/peptide_mass/peptide-mass.pl?RND454178\">PeptideMass</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protparam/protparam.cgi?RND454178\">ProtParam</a>\n",
      "<br>\n",
      "<a href=\"/cgi-bin/protscale/protscale.pl?RND454178\">ProtScale</a>\n",
      "</td>\n",
      "</tr>\n",
      "</table>\n",
      "</div>\n",
      "    </main>\n",
      "    <!-- end of main, start of footer -->\n",
      "    <footer class=\"footer-grid\">\n",
      "\t    <div class=\"footer-elmt\">\n",
      "            <img alt=\"SIB logo\" src=\"/images/sib_images/sib-mark.svg\">\n",
      "\t\t</div>\n",
      "\t\t<div class=\"footer-elmt\" property=\"schema:creator\" typeof=\"schema:NGO\">\n",
      "           <a href=\"https://www.expasy.org\" target=\"_blank\">Expasy</a> \n",
      "\t\t   is operated by the <a property=\"schema:url\" href=\"https://sib.swiss\" target=\"_blank\">SIB Swiss Institute of Bioinformatics</a>\n",
      "\t\t   <br>\n",
      "\t\t   <a href=\"https://www.expasy.org/terms-of-use\" target=\"_blank\">Terms of Use</a>\n",
      "           |\n",
      "           <a href=\"https://www.sib.swiss/privacy-policy\" target=\"_blank\">Privacy policy</a>\n",
      "\t    </div>\n",
      "\t    <div class=\"footer-elmt\">\n",
      "\t        <a href=\"#top\">Back to the top</a>\n",
      "\t    </div>\n",
      "    </footer>\n",
      "    <!-- end of footer -->\n",
      "</body>\n",
      "</html>\n",
      "ID   RND_454178              Unreviewed;        50 AA.\n",
      "AC   RND454178;\n",
      "DT   18-Nov-2025.\n",
      "DE   Randomly generated sequence, created by Expasy tool RandSeq.\n",
      "CC   -!- MISCELLANEOUS: This sequence was generated using average amino acid composition.\n",
      "SQ   SEQUENCE 50 AA; 5577 MW; F6D65D5CD1CDBA1B CRC64;\n",
      "     LAFSVYKTIG RLDPNELVWD ISGGHICMAQ PNEKTQAERD SGTLLKRAFV \n",
      "//\n",
      "Saved FASTA as test_randseq_1.fasta\n",
      ">RND454178\n",
      "LAFSVYKTIGRLDPNELVWDISGGHICMAQPNEKTQAERDSGTLLKRAFV\n"
     ]
    }
   ],
   "source": [
    "##Generates one sequence, prints the Swiss-Prot output, converts it to FASTA, and saves the output file.\n",
    "\n",
    "text_output = generate_randseq(50)           # ðŸ”¶ CHANGE THIS (sequence length)\n",
    "print(text_output)\n",
    "\n",
    "fasta_seq = swissprot_to_fasta(text_output, outname=\"test_randseq_1.fasta\")             # ðŸ”¶ CHANGE THIS (output file name)\n",
    "print(fasta_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d80002c-15b0-467a-b928-7316f892f2e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "^^ this produced one 50aa sequence and saved it here: /Users/derbhladuffy/Desktop/diss_project/annaconda_env_colabfold/test_randseq_1.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8cbd112-ddb0-49b0-80ee-45755b5984f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Let's try the exact same again except this time lets do 10 sequences and have them 20 aa long, and have them saved in an output folder called like randseq_expasy_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b69b7ac8-6b81-4008-aa60-3263b77c9c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "def generate_randseq(length=20):                            # ðŸ”¶ 20 AA sequences\n",
    "    url = \"https://web.expasy.org/cgi-bin/randseq/randseq.pl\"\n",
    "    data = {\n",
    "        \"size\": str(length),                                    # ðŸ”¶ This uses the length above\n",
    "        \"composition\": \"average\"   # default mode             # ðŸ”¶ CHANGE IF YOU WANT: \"average\", \"equal\", \"specific\", \"user\"\n",
    "    }\n",
    "\n",
    "    r = requests.post(url, data=data)\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "    seq_block = soup.find(\"pre\")\n",
    "\n",
    "    if not seq_block:\n",
    "        print(\"Could not find sequence in response.\")\n",
    "        return None\n",
    "\n",
    "    return seq_block.text.strip()   # Return Swiss-Prot text\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b3f5e501-4c80-4685-9a69-dc5f38f4f5f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def swissprot_to_fasta(text, outname):\n",
    "    # Split into lines\n",
    "    lines = text.splitlines()\n",
    "\n",
    "    # Extract accession number\n",
    "    ac_line = [l for l in lines if l.startswith(\"AC\")][0]\n",
    "    accession = ac_line.split()[1].replace(\";\", \"\")\n",
    "\n",
    "    # Extract sequence lines between SQ and //\n",
    "    seq_lines = []\n",
    "    capture = False\n",
    "    for line in lines:\n",
    "        if line.startswith(\"SQ\"):\n",
    "            capture = True\n",
    "            continue\n",
    "        if capture:\n",
    "            if line.startswith(\"//\"):\n",
    "                break\n",
    "            seq_lines.append(line.strip())\n",
    "\n",
    "    # Remove spaces and join\n",
    "    sequence = \"\".join(seq_lines).replace(\" \", \"\")\n",
    "\n",
    "    # Build FASTA text\n",
    "    fasta = f\">{accession}\\n{sequence}\"\n",
    "\n",
    "    # Save\n",
    "    with open(outname, \"w\") as f:\n",
    "        f.write(fasta + \"\\n\")\n",
    "\n",
    "    print(f\"Saved FASTA: {outname}\")\n",
    "    return fasta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "71a4f357-b655-455c-a04f-e076e824cd77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Generating sequence 1 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_1.fasta\n",
      "\n",
      "--- Generating sequence 2 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_2.fasta\n",
      "\n",
      "--- Generating sequence 3 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_3.fasta\n",
      "\n",
      "--- Generating sequence 4 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_4.fasta\n",
      "\n",
      "--- Generating sequence 5 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_5.fasta\n",
      "\n",
      "--- Generating sequence 6 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_6.fasta\n",
      "\n",
      "--- Generating sequence 7 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_7.fasta\n",
      "\n",
      "--- Generating sequence 8 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_8.fasta\n",
      "\n",
      "--- Generating sequence 9 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_9.fasta\n",
      "\n",
      "--- Generating sequence 10 ---\n",
      "Saved FASTA: randseq_expasy_output/randseq_10.fasta\n"
     ]
    }
   ],
   "source": [
    "# Create output folder\n",
    "output_folder = \"randseq_expasy_output\"                    # ðŸ”¶ CHANGE THIS: name of output folder\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Generate 10 sequences, each 20 AA\n",
    "for i in range(1, 11):                                   # ðŸ”¶ CHANGE THIS: how many sequences to generate\n",
    "    print(f\"\\n--- Generating sequence {i} ---\")\n",
    "\n",
    "    # Generate Swiss-Prot entry\n",
    "    text_output = generate_randseq(20)                   # ðŸ”¶ CHANGE THIS: AA length for each sequence\n",
    "\n",
    "    # Build filename inside output folder\n",
    "    outname = os.path.join(output_folder, f\"randseq_{i}.fasta\")    # ðŸ”¶ CHANGE THIS IF YOU WANT CUSTOM FILENAMES\n",
    "\n",
    "    # Convert and save\n",
    "    fasta = swissprot_to_fasta(text_output, outname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb72bc76-df33-4816-afd0-f2f0c69e1d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "## this produuced 10 separete fasta files... now let's merge this into one!..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1d2fdae7-8f1e-4092-86d5-64758196fad7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged FASTA saved as: randseq_expasy_output/randseq_merged.fasta\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "output_folder = \"randseq_expasy_output\"                    # ðŸ”¶ CHANGE THIS if your folder name changes\n",
    "merged_filename = os.path.join(output_folder, \"randseq_merged.fasta\")\n",
    "\n",
    "with open(merged_filename, \"w\") as outfile:\n",
    "    for fname in sorted(os.listdir(output_folder)):\n",
    "        if fname.endswith(\".fasta\"):\n",
    "            file_path = os.path.join(output_folder, fname)\n",
    "            with open(file_path) as infile:\n",
    "                # Strip trailing newlines but DO NOT add extra blank lines\n",
    "                outfile.write(infile.read().strip() + \"\\n\")\n",
    "                \n",
    "print(f\"Merged FASTA saved as: {merged_filename}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa11edfb-df81-4d8d-8bc4-f256589777ac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (py310_env)",
   "language": "python",
   "name": "py310_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
